### PR TITLE
Properly handle the name query param in create issue attachment API

### LINKED ISSUE #30766: Cannot create issue attachment via the API when using query parameter
### Description

In previous versions I used to be able to create issue attachments via the API like so (as per the docs):

```
curl -X 'POST' "http://host/api/v1/repos/$org/$repo/issues/$index/assets?name=Test" \
  -H "Accept: application/json" \
  -H "Authorization: token $TOKEN" \
  -H "Content-Type: multipart/form-data" \
  -F "attachment=@test.jpg;type=image/jpeg"
```

However now the server responds:

> "This file extension or type is not allowed to be uploaded."

The attachment is created fine if the name query parameter is removed.

### Gitea Version

1.21.11

### Can you reproduce the bug on the Gitea demo site?

Yes

### Log Gist

_No response_

### Screenshots

_No response_

### Git Version

_No response_

### Operating System

_No response_

### How are you running Gitea?

Binary download, systemd

### Database

None

### PR DESCRIPTION
**Pull request by kemzeb**
Fixes #30766 by passing both the original file name and an attachment name to the lower API calls and performing verification on just the original file name. This behavior is pretty much in-line with how we handle attachment names in `EditIssueAttachment()`, where we don't perform any verification against the new name and just go straight to the DB to change it.

### CODE REVIEW

File: services/attachment/attachment.go [hunk 1]
```diff
@@ -50,3 +50,17 @@ func UploadAttachment(ctx context.Context, file io.Reader, allowedTypes string,
 
 	return NewAttachment(ctx, attach, io.MultiReader(bytes.NewReader(buf), file), fileSize)
 }
+
+// UploadAttachmentWithCustomizableName is like UploadAttachment, but you pass both the original file name and the attachment name
+// (which could be the same).
+func UploadAttachmentWithCustomizableName(ctx context.Context, file io.Reader, allowedTypes string, fileSize int64, originalFileName string, attach *repo_model.Attachment) (*repo_model.Attachment, error) {
```
Comment (kemzeb) [inline services/attachment/attachment.go]: This function has almost the exact same definition as `UploadAttachment()`. I would need to tweak a couple areas if I were to change its signature so if that's alright to do let me know!

Comment (lunny) [reply services/attachment/attachment.go]: I think we should change `UploadAttachment` and fix all the places but not only this one.

**Comment by kemzeb** [comment]
An example using the swagger API:
```console
# URL
http://localhost:3000/api/v1/repos/korvax/test-repo/issues/1/assets?name=MyNameIsNobody

# Response
{
  "id": 5,
  "name": "MyNameIsNobody",
  "size": 12,
  "download_count": 0,
  "created_at": "2024-04-30T07:32:31Z",
  "uuid": "132694fe-f70c-428e-82d1-66edc88a80b2",
  "browser_download_url": "http://localhost:3000/attachments/132694fe-f70c-428e-82d1-66edc88a80b2"
}
```

File: routers/api/v1/repo/release_attachment.go [hunk 2]
```diff
@@ -219,23 +219,23 @@ func CreateReleaseAttachment(ctx *context.APIContext) {
...
-	if filename == "" {
+	if attachmentName == "" {
 		ctx.Error(http.StatusBadRequest, "CreateReleaseAttachment", "Could not determine name of attachment.")
 		return
 	}
 
 	// Create a new attachment and save the file
-	attach, err := attachment.UploadAttachment(ctx, content, setting.Repository.Release.AllowedTypes, size, &repo_model.Attachment{
-		Name:       filename,
+	attach, err := attachment.UploadAttachment(ctx, content, setting.Repository.Release.AllowedTypes, size, attachmentName, &repo_model.Attachment{
```
Comment (kemzeb) [inline routers/api/v1/repo/release_attachment.go]: The original file name string is hiding within an if block that I can't reach. The original behavior shouldn't have changed, but we may encounter the same problem as the issue we're trying to fix had

File: services/attachment/attachment.go (hunk 1 above)
Comment (kemzeb) [reply services/attachment/attachment.go]: Changed `UploadAttachment()` and updated the affected areas 👍

File: routers/api/v1/repo/release_attachment.go (hunk 2 above)
Comment (lunny) [reply routers/api/v1/repo/release_attachment.go]: I think we can still also have a variable named `filename` and assign it like you did for `attachmentName`

**Comment by lunny** [comment]
Maybe we should have a filename on the `Attachment` struct as a final resolution?

File: routers/api/v1/repo/release_attachment.go (hunk 2 above)
Comment (kemzeb) [reply routers/api/v1/repo/release_attachment.go]: Done, and I fallback to assigning `attachmentName` to it in the `else` block

**Review by KN4CK3R — CHANGES_REQUESTED** [review]
Now you can bypass the extension check by uploading `file.good` with `?name=file.bad`.

**Comment by kemzeb** [comment]
> Now you can bypass the extension check by uploading `file.good` with `?name=file.bad`.

Ah, I think I might not have understood well enough/was confused why we looked at file extensions (when we already performed MIME sniffing). Now, it looks to me like we perform these extension checks to reject extensions that may look malicious to users or otherwise undesirable (e.g. file.exe).

If this is the case, would it be alright if I close this PR and open a new one that performs a verification check when users edit attachment names? Thank you both for the review!

**Comment by kemzeb** [comment]
One other thing we could do is determine if an attachment without an extension should be allowed (since this seems to be the problem the reporter is facing specifically).

**Comment by kemzeb** [comment]
Going to close this as bypassing the file extension check will surprise admins since they expect either the default extensions we provide or their own to be honored.

**Pull request closed without merging**